framework/client.py

```diff
@@ -121,36 +139,40 @@ def isCommunication():
       return packet["type"] == "simulation-communicate"
 
     if isError():
-      print(packet["content"])
-      #print(packet["content"]["reason"])
+      self.failsafe(packet["content"])
     elif isInitialParams():
       self.frameworkID = packet["content"]["frameworkID"]
-      self.handleSimulationStart(packet["content"])
-    elif isClose():
-      self.handleSimulationStop(packet["content"])
-      self.alive = False
-      #The connection is officialy dead we need to terminate the handling loop,
-      #to achieve this we populate the message queue with a confirmation packet
-      packet = {'type': 'simulation-close', 'content': {'simulationID': self.simulationID}}
-      asyncio.run_coroutine_threadsafe(self.messageQueue.put(packet),
-      loop)
+      self.attempt(self.handleSimulationStart, packet)
+      self.endSimulation()
     elif isUpdate():
-      self.handleSimulationDataUpdate(packet["content"])
+      self.attempt(self.handleSimulationDataUpdate, packet)
     elif isCommunication():
       print("\n\nReceived go-ahead at ", time.time())
-      self.handleSimulationCommunication(packet["content"])
+      self.attempt(self.handleSimulationCommunication, packet)
       self.shouldAwait = False
-
-  def synchronize(self, sleepTime=1):
+    elif isClose():
+      self.attempt(self.handleSimulationStop, packet)
+      self.endSimulation()
+
+  def endSimulation(self):
+    #At this point the actual algorithm must have made sure it will terminate
+    self.alive = False
+    #The connection is officialy dead we need to terminate the handling loop,
+    #to achieve this we populate the message queue with a confirmation packet
+    packet = {'type': 'simulation-close', 'content': {'simulationID': self.simulationID}}
+    asyncio.run_coroutine_threadsafe(self.messageQueue.put(json.dumps(packet)),
+    loop)
+
+  def synchronize(self, sleepTime):
     self.shouldAwait = True
-    while (self.shouldAwait):
+    while (self.shouldAwait): #TODO: Is there a better way to freeze this thread?
       time.sleep(sleepTime)
```

Comment:
For now, the current method is necessary to force a sleep before the message is sent. I'll replace it with locks or something afterwards.